app/models/agents/gap_detector_agent.rb

```diff
@@ -42,26 +47,89 @@ def working?
 
     def receive(incoming_events)
       incoming_events.sort_by(&:created_at).each do |event|
-        memory['newest_event_created_at'] ||= 0
-
         if !interpolated['value_path'].present? || Utils.value_at(event.payload, interpolated['value_path']).present?
-          if event.created_at.to_i > memory['newest_event_created_at']
-            memory['newest_event_created_at'] = event.created_at.to_i
-            memory.delete('alerted_at')
+          stream = nil
+          if stream_path
+            stream = Utils.value_at(event.payload, stream_path)
+            next unless stream
+          end
+          storage = event_storage(stream)
+          if event.created_at.to_i > storage.newest_event_created_at
+            storage.newest_event_created_at = event.created_at.to_i
+            storage.clear_alerted_at
           end
         end
       end
     end
 
     def check
       window = interpolated['window_duration_in_days'].to_f.days.ago
-      if memory['newest_event_created_at'].present? && Time.at(memory['newest_event_created_at']) < window
-        unless memory['alerted_at']
-          memory['alerted_at'] = Time.now.to_i
-          create_event payload: { message: interpolated['message'],
-                                  gap_started_at: memory['newest_event_created_at'] }
+      if stream_path
+        memory.each_key do |stream|
+          next unless memory[stream].is_a? Hash
+          next unless memory[stream].has_key? 'newest_event_created_at'
+          check_stream(window, stream)
+        end
+      else
+        check_stream(window)
+      end
+    end
+
+    private
+
+    def check_stream(window, stream=nil)
+      storage = event_storage(stream)
+      if storage.newest_event_created_at? && Time.at(storage.newest_event_created_at) < window
+        unless storage.alerted_at
```

Comment:
Inverting the conditionals would allow you to remove the indenation:
```
return if !storage.newest_event_created_at? || Time.at(storage.newest_event_created_at) > window
return if storage.alerted_at
```